### PR TITLE
Upgrade to react-scripts 4

### PR DESCRIPTION
**Pull request by pabloferro**
## Summary

Upgrade to react-scripts 4.

### Pending
- Move eslint rules to the corresponding packages

### CODE REVIEW

File: generators/app/templates/gitignore [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+# production
+/build
+
+# misc
+.DS_Store
+.env.local
+.env.development.local
+.env.test.local
+.env.production.local
```
Comment (alejobh) [inline generators/app/templates/gitignore]: i know its WIP but for the future

```
.env.*
```

Comment (pabloferro) [reply generators/app/templates/gitignore]: Nice suggestion!

File: .eslintrc.js [hunk 2]
```diff
@@ -1,14 +1,44 @@
...
+      }
+    ],
+    "@typescript-eslint/naming-convention": [
+      "error",
+      {
+        "selector": "default",
+        "format": ["camelCase"]
+      },
+      {
+        "selector": "variableLike",
```
Comment (damfinkel) [inline .eslintrc.js:25]: With this we allow functions to be UPPER_CASE, isn't it better to use:
```
      {
        "selector": ["variable", "parameter"],
        "format": ["camelCase", "UPPER_CASE", "PascalCase"]
      },
      {
        "selector": ["typeLike", "function"],
        "format": ["PascalCase"]
      },
```

File: README.md [hunk 3]
```diff
@@ -101,13 +101,11 @@ Additional to clean libraries are added
...
 |[react-router-dom](https://github.com/ReactTraining/react-router/tree/master/packages/react-router-dom)| DOM bindings for React Router.
 |[react-router](https://github.com/ReactTraining/react-router)| Routing for React.
-|[seamless-immutable](https://github.com/rtfeldman/seamless-immutable)| Immutable data structures for JavaScript.
 |[history](https://www.npmjs.com/package/history)| Manage session history anywhere JavaScript runs.
 |[i18next](https://www.i18next.com/)| An internationalization-framework written in and for JavaScript.
 |[react-spinkit](https://kyleamathews.github.io/react-spinkit/)| Spinners library.
 |[typescript](https://www.typescriptlang.org/)| Brings you optional static type-checking along with the latest ECMAScript
 feature.
-|[seamless-immutable](https://github.com/rtfeldman/seamless-immutable)| Immutable data structures for JavaScript.
 |[babel-module-resolver](https://github.com/tleunen/babel-plugin-module-resolver)| Custom module resolver plugin for Babel.
```
Comment (damfinkel) [inline README.md]: Shouldn't we remove this too?

File: generators/app/constants.js [hunk 4]
```diff
@@ -90,13 +107,16 @@ module.exports.BOOTSTRAP_TYPES = [
 ];
 
 module.exports.DEV_DEPENDENCIES = [
-  'eslint-plugin-import@2.18.2',
+  'eslint-plugin-import@2.22.1',
   'eslint-plugin-jsx-a11y@6.2.3',
   'eslint-plugin-prettier@3.1.0',
   'eslint-plugin-react@7.14.3',
+  // TODO: Do we need this?
```
Comment (damfinkel) [inline generators/app/constants.js]: https://github.com/Wolox/eslint-config-wolox-react/blob/master/index.js

It's being used, I think it as something related to optional chaining. In any case, to remove it we'd have to change the linter

File: generators/app/constants.js [hunk 5]
```diff
@@ -126,9 +143,9 @@ module.exports.DEV_DEPENDENCIES = [
 ];
 
 module.exports.DEPENDENCIES = [
-  'postcss@^7.0.7',
-  'react@^16.6.3',
-  'react-dom@^16.6.3',
+  // 'postcss@^7.0.7',
```
Comment (damfinkel) [inline generators/app/constants.js]: Why is this commented?

File: generators/app/index.js [hunk 6]
```diff
@@ -70,24 +70,32 @@ class GeneratorReact extends Generator {
...
-      this.spawnCommand('npm', ['install']).on('close', () => resolve());
+      this.spawnCommand('npm', ['install'], { cwd: `${process.cwd()}/${this.projectName}` }).on('close', () =>
+        resolve()
+      );
     });
   }
 
   async end() {
     this.log('Ending...');
+    await linterAutofix
```
Comment (damfinkel) [inline generators/app/index.js:90]: Why are we ignoring?

File: generators/app/steps/CustomizableSteps/index.js [hunk 7]
```diff
@@ -1,15 +1,14 @@
...
 
-const copyAllFiles = require('./copyFiles');
+const copyAllFiles = require('./copyAllFiles');
 const { DEPENDENCIES, DEV_DEPENDENCIES } = require('./constants');
-const { CUSTOMIZED_PROMPTS } = require('./prompts');
 
 const CustomizableGenerator = {
   prompting() {
-    return this.prompt(CUSTOMIZED_PROMPTS);
+    return Promise.resolve();
```
Comment (damfinkel) [inline generators/app/steps/CustomizableSteps/index.js:8]: You removed the prompts? Why?

File: generators/app/tasks/createReactApp.js [hunk 8]
```diff
@@ -2,7 +2,7 @@ const runCommand = require('./runCommand');
 
 module.exports.installCRA = function installCRA() {
   return runCommand({
-    command: ['npm', ['install', '--global', 'create-react-app@^3.0.0']],
+    command: ['npm', ['install', '--global', 'create-react-app']],
```
Comment (damfinkel) [inline generators/app/tasks/createReactApp.js]: SHould be hardcode to CRA 4 to avoid compatibility problems?

File: generators/app/templates/.babelrc.ejs [hunk 9]
```diff
@@ -1,22 +1,2 @@
...
-          "~utils": "./src/utils",
-          "~app": "./src/app",
-          "~hooks": "./src/app/hooks",
-          "~contexts": "./src/app/contexts"
-        }
-      }
-    ]<% } %>
-  ]
-}
+// TODO: Remove?
```
Comment (damfinkel) [inline generators/app/templates/.babelrc.ejs]: I think so

File: generators/app/templates/eslintrc.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+      }
+    ],
+    '@typescript-eslint/naming-convention': [
+      'error',
+      {
+        selector: 'default',
+        format: ['camelCase']
+      },
+      {
+        selector: 'variableLike',
```
Comment (damfinkel) [inline generators/app/templates/eslintrc.js]: same as before here

File: generators/app/templates/src/components/ProviderWrapper/index.tsx [hunk 11]
```diff
@@ -5,13 +5,13 @@ interface ActionType {
 }
 
 interface Props<U extends {}, V> {
-  Context: React.Context<any>;
+  context: React.Context<any>;
```
Comment (damfinkel) [inline generators/app/templates/src/components/ProviderWrapper/index.tsx]: I just changed this to `React.Context<{ state: U, dispatch: React.Dispatch<V> }>;` in the cookbook

File: generators/app/templates/src/components/Spinner/components/loading.tsx [hunk 12]
```diff
@@ -1,4 +1,5 @@
 import React from 'react';
+// eslint-disable-next-line import/no-named-as-default
```
Comment (damfinkel) [inline generators/app/templates/src/components/Spinner/components/loading.tsx:2]: why is this needed?

File: generators/app/templates/src/components/Spinner/index.tsx [hunk 13]
```diff
@@ -13,7 +13,7 @@ interface SpinnerConfig extends SpinnerProps {
   classNameContainer?: string;
 }
 
-export const withSpinner = (spinnerConfig: SpinnerConfig = {}) => <P extends object>(
+export const withSpinner = (spinnerConfig: SpinnerConfig = {}) => <P extends {}>(
```
Comment (damfinkel) [inline generators/app/templates/src/components/Spinner/index.tsx:18]: Isn't this against one of the typescript rules? the {} I mean

File: generators/app/templates/src/config/context.ts [hunk 14]
```diff
@@ -12,7 +12,7 @@ export const contextFactory = <State, Action>(initialState: State) => {
     dispatch: () => {}
   });
 
-  const useSelector = (selector: (arg: State) => any) => {
+  const useSelector = <T>(selector: (arg: State) => T) => {
```
Comment (damfinkel) [inline generators/app/templates/src/config/context.ts:15]: Nice

File: generators/app/templates/src/constants/README.md [hunk 15]
```diff
@@ -0,0 +1,3 @@
+# Constants
```
Comment (damfinkel) [inline generators/app/templates/src/constants/README.md:1]: What's this?

File: generators/app/templates/src/index.ejs [hunk 16]
```diff
@@ -1,28 +1,36 @@
 import React from 'react';
-import ReactDOM from 'react-dom';
-import { AppContainer } from 'react-hot-loader';
+import { render } from 'react-dom';
+// import { AppContainer } from 'react-hot-loader';
```
Comment (damfinkel) [inline generators/app/templates/src/index.ejs]: remove?

File: generators/app/templates/src/index.ejs [hunk 17]
```diff
@@ -1,28 +1,36 @@
...
 };
 
 // Render once
-render();
+renderApp();
 
 // Webpack Hot Module Replacement API
 if (module.hot) {<% if (features && features.customized) { %>
-  module.hot.accept('./app', () => {<% } else {%>
+  module.hot.accept('./components/App', () => {<% } else {%>
```
Comment (damfinkel) [inline generators/app/templates/src/index.ejs]: shouldn't we remove this?

File: generators/app/templates/src/utils/displayName.ts [hunk 18]
```diff
@@ -1,2 +1,4 @@
-export const getDisplayName = <P>(WrappedComponent: React.ComponentType<P>) =>
+import { ComponentType } from 'react';
+
+export const getDisplayName = <P>(WrappedComponent: ComponentType<P>) =>
```
Comment (damfinkel) [inline generators/app/templates/src/utils/displayName.ts:3]: Are w using this?

**Review by damfinkel — CHANGES_REQUESTED** [review]

File: generators/app/constants.js (hunk 5 above)
Comment (pabloferro) [reply generators/app/constants.js]: Because cra already installs them.

File: generators/app/index.js (hunk 6 above)
Comment (pabloferro) [reply generators/app/index.js:90]: It was reporting the error two times, also finishing the bootstrap on that step.
This way you'll have an error but the rest will continue.

File: generators/app/steps/CustomizableSteps/index.js (hunk 7 above)
Comment (pabloferro) [reply generators/app/steps/CustomizableSteps/index.js:8]: They were optional components that we don't want anymore. I don't think we have a case for prompts right now. We can restore it if needed.

File: generators/app/tasks/createReactApp.js (hunk 8 above)
Comment (pabloferro) [reply generators/app/tasks/createReactApp.js]: Yeah, good catch

File: generators/app/templates/src/components/Spinner/components/loading.tsx (hunk 12 above)
Comment (pabloferro) [reply generators/app/templates/src/components/Spinner/components/loading.tsx:2]: I'm having an issue where 'react-spinkit' has a named export called Spinner and eslint thinks that's an error.

File: generators/app/templates/src/components/Spinner/index.tsx (hunk 13 above)
Comment (pabloferro) [reply generators/app/templates/src/components/Spinner/index.tsx:18]: Yeah, we could ban that and use another alternative.
I left it this way because, it's the way react and other libs type props.
ref: https://github.com/typescript-eslint/typescript-eslint/issues/2063

File: generators/app/templates/src/constants/README.md (hunk 15 above)
Comment (pabloferro) [reply generators/app/templates/src/constants/README.md:1]: A placeholder to commit this empty folder

File: generators/app/templates/src/utils/displayName.ts (hunk 18 above)
Comment (pabloferro) [reply generators/app/templates/src/utils/displayName.ts:3]: In withSpinner

**Review by damfinkel — APPROVED** [review]
